crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -130,22 +203,21 @@ impl ArrayHeap {
         }
         outputs
     }
-
-    pub(crate) fn acir_gen(&self, evaluator: &mut Evaluator, array_id: ArrayId, array_len: u32) {
+    pub(crate) fn generate_permutation_constraints(
+        &self,
+        evaluator: &mut Evaluator,
+        array_id: ArrayId,
+        array_len: u32,
+    ) {
         let (len, read_write) = match self.typ {
             ArrayType::Init(_, _) | ArrayType::WriteOnly => (0, true),
             ArrayType::ReadOnly(last) => (last.unwrap_or(self.trace.len()), false),
             ArrayType::ReadWrite(last) => (last.unwrap_or(self.trace.len()), true),
         };
-
         if len == 0 {
             return;
         }
-        evaluator.opcodes.push(AcirOpcode::Block(MemoryBlock {
-            id: AcirBlockId(array_id.as_u32()),
-            len: array_len,
-            trace: self.trace.clone(),
-        }));
+        self.add_block_opcode(evaluator, array_id, array_len);
```

Comment:
We're creating a `Block` opcode here in a function which only gets entered if the backend has specified that it doesn't support `Block` opcodes. Why is this?

Reply:
It is needed anyway for acvm to solve witness, we do not generate constraints if the backend supports the opcode.

Reply:
Ok, I think we should add some documentation here about this and how these permutation constraints prevent us from needing to pass the block opcode to the backend to lay down constraints.